protocol/signalfx/trace_jaeger.go

```diff
@@ -0,0 +1,316 @@
+		if remote == nil {
+			remote = &trace.Endpoint{}
+		}
+	}
+
+	for i := range s.Tags {
+		switch s.Tags[i].Key {
+		case string(ext.PeerHostIPv4):
+			ip := convertPeerIPv4(s.Tags[i])
+			if ip == "" {
+				continue
+			}
+			ensureRemote()
+			remote.Ipv4 = pointer.String(ip)
+		// ipv6 host is always string
+		case string(ext.PeerHostIPv6):
+			if s.Tags[i].VStr != nil {
+				ensureRemote()
+				remote.Ipv6 = s.Tags[i].VStr
+			}
+		case string(ext.PeerPort):
+			port := convertPeerPort(s.Tags[i])
+			if port == 0 {
+				continue
+			}
+			ensureRemote()
+			remote.Port = &port
+		case string(ext.PeerService):
+			ensureRemote()
+			remote.ServiceName = s.Tags[i].VStr
+		case string(ext.SpanKind):
+			kind = convertKind(s.Tags[i])
+		default:
+			val := tagValueToString(s.Tags[i])
+			if val != "" {
+				tags[s.Tags[i].Key] = val
+			}
+		}
+	}
+	return kind, remote, tags
+}
+
+func convertKind(tag *jThrift.Tag) *string {
+	var kind *string
+	switch tag.GetVStr() {
+	case string(ext.SpanKindRPCClientEnum):
+		kind = &ClientKind
+	case string(ext.SpanKindRPCServerEnum):
+		kind = &ServerKind
+	case string(ext.SpanKindProducerEnum):
+		kind = &ProducerKind
+	case string(ext.SpanKindConsumerEnum):
+		kind = &ConsumerKind
+	}
+	return kind
+}
+
+func convertPeerIPv4(tag *jThrift.Tag) string {
+	switch tag.VType {
+	case jThrift.TagType_STRING:
+		if ip := net.ParseIP(tag.GetVStr()); ip != nil {
+			return ip.To4().String()
+		}
+	case jThrift.TagType_LONG:
+		localIP := make(net.IP, 4)
+		binary.BigEndian.PutUint32(localIP, uint32(tag.GetVLong()))
+		return localIP.String()
+	}
+	return ""
+}
+
+func convertPeerPort(tag *jThrift.Tag) int32 {
+	switch tag.VType {
+	case jThrift.TagType_STRING:
+		if port, err := strconv.ParseUint(tag.GetVStr(), 10, 16); err == nil {
+			return int32(port)
+		}
+	case jThrift.TagType_LONG:
+		return int32(tag.GetVLong())
+	}
+	return 0
+}
+
+// materializeWithJSON converts log Fields into JSON string, or just the field
+// value of the event field, if present.
+func materializeWithJSON(logFields []*jThrift.Tag) ([]byte, error) {
+	fields := make(map[string]string, len(logFields))
+	for i := range logFields {
+		fields[logFields[i].Key] = tagValueToString(logFields[i])
+	}
+	if event, ok := fields["event"]; ok && len(fields) == 1 {
+		return []byte(event), nil
+	}
+	return json.Marshal(fields)
+}
+
+var pads = []string{
+	"",
+	"0",
+	"00",
+	"000",
+	"0000",
+	"00000",
+	"000000",
+	"0000000",
+	"00000000",
+	"000000000",
+	"0000000000",
+	"00000000000",
+	"000000000000",
+	"0000000000000",
+	"00000000000000",
+	"000000000000000",
+}
+
+// The way IDs get converted to strings in some of the jaeger code, leading 0s
+// can be dropped, which will cause the ids to fail validation on our backend.
+func padID(id string) string {
+	expectedLen := 0
+	if len(id) < 16 {
+		expectedLen = 16
+	} else if len(id) > 16 && len(id) < 32 {
+		expectedLen = 32
+	} else {
+		return id
+	}
+
+	return pads[expectedLen-len(id)] + id
```

Comment:
this is more efficient than fmt.Sprintf( "%016s", s) ? i feel likely it's not.

Reply:
Benchmarks say it is more efficient with preallocated strings and direct concatenation: https://gist.github.com/keitwb/fa3f199b27061672ff31a17d4b69bfbe